# Strelka2 Somatic source code analysis 


### Handles input read alignments
```C++ 
processInputReadAlignment
// filter read 
    starling_read_filter_shared
    normalizeAlignment
        collapseAlignmentIndels
            collapseInsert //2M1I2D2M -> 2M1D2M
        leftShiftAlignmentIndels 
            leftShiftIndel
        normalizeEdgeIndels 
            leftEdgeIndelCollapse
        posProcessor.insert_read
            rbuff.add_read_alignment
            init_read_segment
                    init_read_segment
                        addAlignmentIndelsToPosProcessor
                            process_edge_insert
                            process_edge_delete
                            process_swap
                            process_simple_indel 
                                // 
                                finish_indel_sppr (insert_indel)
                                    addIndelObservation
                                        // get information used to summarize indel output
                                        getAlleleReportInfo
                                        // Initialize error rates
                                        initializeAuxInfo
                                        // insert indel error rate 
                                        addIndelObservation
```

## Stage PIPELINE
### Stage 0 HEAD
HEAD (0)

```c++
initializeSplicedReadSegmentsAtPos
    init_read_segment
        addAlignmentIndelsToPosProcessor
            process_edge_insert
            process_edge_delete
            process_swap
            process_simple_indel 
                // 
                finish_indel_sppr (insert_indel)
                    addIndelObservation
                        // get information used to summarize indel output
                        getAlleleReportInfo
                        // Initialize error rates
                        initializeAuxInfo
                        // insert indel error rate 
                        addIndelObservation
                    
```
### Stage 1 READBUFFER
HEAD -> READBUFFER (499)  
READ_BUFFER is where most normal genomic reads are read in and processed for indels, it needs enough room to collect the full candidate indel map before we start read alignment

```c++
// For all reads buffered at the current position:
// 1) determine the set of candidate indels that the read overlaps
// 2) determine the set of private indels within the read's discovery alignments
// 3) Find the most likely alignment given both sets of indels
// 4) evaluate the probability that the read supports each candidate indel vs. the reference
align_pos(pos)
    realignAndScoreRead
        getCandidateAlignments
        add_indels_in_range
// Get the keys of the indels present in the candidate alignment  
        getAlignmentIndels
//if true recompute 
        make_start_pos_alignment
//  to prevent truncated search, we must put all non-present remove-only indels last in the order list:
        sort_remove_only_indels_last
// NOTE Recursively build potential alignment paths and push them into the candidate alignment set
        candidate_alignment_search
// Find the most likely alignment and most likely alignment for each indel state for every indel in indel_status_map
        scoreCandidateAlignmentsAndIndels
            scoreCandidateAlignments
                scoreCandidateAlignment
                finishRealignment
            score_indels
                late_indel_normalization_filter

pileup_pos_reads(pos)
    pileup_read_segment
        create_mismatch_filter_map
        base_call
        // 
        insert_pos_basecall

write_reads(pos)

```

### Stage 2 POST_ALIGN

READBUFFER -> POST_ALIGN (499 + 49 = 548)

POST_ALIGN_STAGE - the end of this stage is where snp and indel
calling occur. It needs to be separated from the end of
READ_BUFFER_STAGE with enough room to allow for the repositioning
of reads to the 'left' during re-alignment.
```c++
process_pos_variants
    // if true 
    process_pos_stats
        isCandidateIndel
            isCandidateIndelImpl
                isCandidateIndelImplTest
    process_pos_sample_stats

    process_pos_variants_impl 
        process_pos_snp_somatic ( pos )
            position_somatic_snv_call
// Calculates somatic variant quality scores
                calculate_result_set_grid  
            write_vcf_somatic_snv_genotype_strand_grid
                get_scoring_features
                updateAlleleEVSScore
        process_pos_indel_somatic(pos);
            isCandidateIndel
            getSingleIndelAlleleVcfSummaryStrings
            get_somatic_indel
                // compute read likelihood 
                calculate_result_set_grid
                // output
                getAlleleSampleReportInfo(normal)
                getAlleleSampleReportInfo(tumor)
                _indelWriter.cacheIndel
    post_align_clear_pos
```


### Stage 3 CLEAR_SITE_ANNOTATION
POST_ALIGN -> CLEAR_SITE_ANNOTATION(548 + 49)
```c++
for (unsigned sampleIndex(0); sampleIndex < sampleCount; ++sampleIndex) {
    sample(sampleIndex).readBuffer.clear_to_pos(pos);
}
```


### Stage 4 CLEAR_READ_BUFFER
POST_ALIGN -> CLEAR_READ_BUFFER (548 + 0)
```C++
for (unsigned sampleIndex(0); sampleIndex < sampleCount; ++sampleIndex)
{
    sample(sampleIndex).readBuffer.clear_to_pos(pos);
}
```
### Stage 5 CLEAR_INDEL_BUFFER
POST_ALIGN -> CLEAR_INDEL_BUFFER (548 + 49 = 597)
```c++
getIndelBuffer().clearIndelsAtPosition(pos);
```
### Stage 6 POST_CALL
POST_ALIGN -> POST_CALL (548 + 1 = 549)
```c++  
for (unsigned sampleIndex(0); sampleIndex < sampleCount; ++sampleIndex) {
    sample_info& sif(sample(sampleIndex));
    sif.estdepth_buff.clear_pos(pos);
    sif.estdepth_buff_tier2.clear_pos(pos);
    // if we are doing short-range phasing, suspend pileup buffer clear
    // while phasing block is being built:
    if (! is_save_pileup_buffer())
    {
        sif.basecallBuffer.clear_to_pos(pos);
    }
}
``` 
### Stage 7 SIZE
POST_ALIGN -> SIZE (548 + 50 = 598)
output compute EVS for Indel and output Indel data
```c++
run_post_call_step
    addIndelWindowData
        writeSomaticIndelVcfGrid
            calculateScoringFeatures
            updateAlleleEVSScore
            // ouput vcf
```